PowerShell 7.4.0-preview.4 fails to start on Windows PC when disconnected from domain network

### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

1. Installed pwsh 7.4.0-preview-4 on a domain joined PC, but the PC was on an external network provided by a mobile hotspot.
2. Tried to launch pwsh-preview after installing


### Expected behavior

```console
PowerShell launches normally and prompt provided.
```


### Actual behavior

```console
PowerShell fails to start with the following message:

The shell cannot be started. A  failure occurred during initialization:
Object reference not set to an instance of an object.
```

Note that if the PC is on the domain network, the preview release starts normally. It does not start if either a different network connection is used that doesn't have access to the domain network, or the network is turned off all together.

I also tested on pwsh 7.3.2 and it works in either state.

I was not able to inspect the exception in the release version as the shell does not stay running. I built a debug version from my working area (that is between preview.3 and preview.4 in terms of code), and found it asserted on the machine in question with the following error:
```
pathToAdd should not be null according to contract of the function.
```

This is being triggered in `GetModulePath` of `ModuleIntrinsics.cs` when it tries to add a path to the module path, but the value it's trying to add is null. 

I tried checking if each of `personalModulePathToUse`, `sharedModulePath` and `systemModulePathToUse` is null before allowing them to be added, and with that change, the debug version of pwsh starts on the machine in question when in the disconnected state, but I'm not clear on what this code is doing with the path indices after adding them so I'm not confident to provide a PR for this.
```


### Error details

_No response_

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.0-preview.4
PSEdition                      Core
GitCommitId                    7.4.0-preview.4
OS                             Microsoft Windows 10.0.19044
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```


### Visuals

_No response_

@iSazonov - what's the trick to building pwsh for the first time?  Following [these directions](https://github.com/PowerShell/PowerShell/blob/master/docs/building/windows-core.md) I installed the Dot NET 8-preview 5 SDK and `Import-Module ./build.psm1; Start-PSBootstrap` runs ok.  I just see a wall of red errors regarding certificate expiration when running `Start-PSBuild` which fails. I also tried .Net SDK v6 and v7.

eg.
```
C:\Users\user\Github\PowerShell\src\Microsoft.PowerShell.Commands.Utility\Microsoft.PowerShell.Commands.Utility.csproj : error NU3037: Package 'Microsoft.Management.Infrastructure 2.0.0' from source 'https://
api.nuget.org/v3/index.json': The author primary signature validity period has expired. [C:\Users\user\Github\PowerShell\src\powershell-win-core\powershell-win-core.csproj]
C:\Users\user\Github\PowerShell\src\Microsoft.PowerShell.Security\Microsoft.PowerShell.Security.csproj : error NU3037: Package 'Microsoft.Management.Infrastructure.Runtime.Unix 2.0.0' from source 'https://api
.nuget.org/v3/index.json': The repository countersignature validity period has expired. [C:\Users\user\Github\PowerShell\src\powershell-win-core\powershell-win-core.csproj]
C:\Users\user\Github\PowerShell\src\Microsoft.PowerShell.Security\Microsoft.PowerShell.Security.csproj : error NU3028: Package 'Microsoft.Management.Infrastructure.Runtime.Unix 2.0.0' from source 'https://api
.nuget.org/v3/index.json': The repository countersignature's timestamping certificate is not trusted by the trust provider. [C:\Users\user\Github\PowerShell\src\powershell-win-core\powershell-win-core.csproj]
C:\Users\user\Github\PowerShell\src\Microsoft.PowerShell.Commands.Utility\Microsoft.PowerShell.Commands.Utility.csproj : error NU3028: Package 'Microsoft.Management.Infrastructure 2.0.0' from source 'https://
api.nuget.org/v3/index.json': The author primary signature's timestamping certificate is not trusted by the trust provider. [C:\Users\user\Github\PowerShell\src\powershell-win-core\powershell-win-core.csproj]
```

--- a/src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs
+++ b/src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs
@@ -967,7 +967,7 @@ namespace System.Management.Automation
 #if UNIX
             return Platform.SelectProductNameForDirectory(Platform.XDG_Type.USER_MODULES);
 #else
-            string myDocumentsPath = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
+            string myDocumentsPath = InternalTestHooks.SetMyDocumentsSpecialFolderToBlank ? string.Empty : Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
             return string.IsNullOrEmpty(myDocumentsPath) ? null : Path.Combine(myDocumentsPath, Utils.ModuleDirectory);
 #endif
         }

--- a/src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs
+++ b/src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs
@@ -1259,26 +1259,30 @@ namespace System.Management.Automation
                 // personalModulePath
                 // sharedModulePath
                 // systemModulePath
-                currentProcessModulePath = AddToPath(currentProcessModulePath, personalModulePathToUse, 0);
-                int insertIndex = currentProcessModulePath.IndexOf(Path.PathSeparator, PathContainsSubstring(currentProcessModulePath, personalModulePathToUse));
-                if (insertIndex != -1)
-                {
-                    // advance past the path separator
-                    insertIndex++;
-                }
 
-                currentProcessModulePath = AddToPath(currentProcessModulePath, sharedModulePath, insertIndex);
-                insertIndex = currentProcessModulePath.IndexOf(Path.PathSeparator, PathContainsSubstring(currentProcessModulePath, sharedModulePath));
-                if (insertIndex != -1)
-                {
-                    // advance past the path separator
-                    insertIndex++;
-                }
+                int insertIndex = 0;
 
-                currentProcessModulePath = AddToPath(currentProcessModulePath, systemModulePathToUse, insertIndex);
+                currentProcessModulePath = UpdatePath(currentProcessModulePath, personalModulePathToUse, ref insertIndex);
+                currentProcessModulePath = UpdatePath(currentProcessModulePath, sharedModulePath, ref insertIndex);
+                currentProcessModulePath = UpdatePath(currentProcessModulePath, systemModulePathToUse, ref insertIndex);
             }
 
             return currentProcessModulePath;
+        }
+
+        private static string UpdatePath(string path, string pathToAdd, ref int insertIndex)
+        {
+            if (!string.IsNullOrEmpty(pathToAdd))
+            {
+                path = AddToPath(path, pathToAdd, insertIndex);
+                insertIndex = path.IndexOf(Path.PathSeparator, PathContainsSubstring(path, pathToAdd));
+                if (insertIndex != -1)
+                {
+                    // advance past the path separator
+                    insertIndex++;
+                }
+            }
+            return path;
         }
 
         /// <summary>

--- a/src/System.Management.Automation/engine/Utils.cs
+++ b/src/System.Management.Automation/engine/Utils.cs
@@ -1559,6 +1559,9 @@ namespace System.Management.Automation.Internal
         internal static bool SetConsoleWidthToZero;
         internal static bool SetConsoleHeightToZero;
 
+        // Simulate 'MyDocuments' returning empty string
+        internal static bool SetMyDocumentsSpecialFolderToBlank;
+
         internal static bool SetDate;
 
         // A location to test PSEdition compatibility functionality for Windows PowerShell modules with

--- a/test/powershell/engine/Module/ModulePath.Tests.ps1
+++ b/test/powershell/engine/Module/ModulePath.Tests.ps1
@@ -247,3 +247,21 @@ Describe "ModuleIntrinsics.GetPSModulePath API tests" -tag @('CI', 'RequireAdmin
         $currentModulePathElements | Should -Contain $mPath
     }
 }
+
+Describe 'ModuleIntrinsics.GetModulePath environment interaction' -tag @('CI') {
+
+    It "GetModulePath does not crash when MyDocuments special folder is empty" {
+        try {
+            [System.Management.Automation.Internal.InternalTestHooks]::SetTestHook('SetMyDocumentsSpecialFolderToBlank', $true)
+            [System.Management.Automation.ModuleIntrinsics]::GetModulePath($null, $null, $null)
+        }
+        finally {
+            [System.Management.Automation.Internal.InternalTestHooks]::SetTestHook('SetMyDocumentsSpecialFolderToBlank', $false)
+        }
+    }
+
+    It "GetModulePath does not crash when MyDocuments special folder is not empty" {
+        [System.Management.Automation.ModuleIntrinsics]::GetModulePath($null, $null, $null)
+    }
+
+}

